crates/sui-graphql-rpc/src/types/query.rs

```diff
@@ -46,9 +54,112 @@ impl Query {
     }
 
     // availableRange - pending impl. on IndexerV2
-    // dryRunTransactionBlock
     // coinMetadata
 
+    /// Simulate running a transaction to inspect its effects without
+    /// committing to them on-chain.
+    ///
+    /// `txBytes` either a `TransactionData` struct or a `TransactionKind`
+    ///     struct, BCS-encoded and then Base64-encoded.  The expected
+    ///     type is controlled by the presence or absence of `txMeta`: If
+    ///     present, `txBytes` is assumed to be a `TransactionKind`, if
+    ///     absent, then `TransactionData`.
+    ///
+    /// `txMeta` the data that is missing from a `TransactionKind` to make
+    ///     a `TransactionData` (sender address and gas information).  All
+    ///     its fields are nullable: `sender` defaults to `0x0`, if
+    ///     `gasObjects` is not present, or is an empty list, it is
+    ///     substituted with a mock Coin object, and `gasPrice` defaults to
+    ///     the reference gas price.
+    ///
+    /// `skipChecks` optional flag to disable the usual verification
+    ///     checks that prevent access to objects that are owned by
+    ///     addresses other than the sender, and calling non-public,
+    ///     non-entry functions.  Defaults to false.
+    ///
+    /// `epoch` the epoch to simulate executing the transaction in.
+    ///     Defaults to the current epoch.
+    async fn dry_run_transaction_block(
+        &self,
+        ctx: &Context<'_>,
+        tx_bytes: String,
+        tx_meta: Option<TransactionMeta>,
+        skip_checks: Option<bool>,
+        epoch: Option<u64>,
+    ) -> Result<DryRunResult> {
+        let skip_checks = skip_checks.unwrap_or(false);
+
+        let sui_sdk_client: &Option<SuiClient> = ctx
+            .data()
+            .map_err(|_| Error::Internal("Unable to fetch Sui SDK client".to_string()))
+            .extend()?;
+        let sui_sdk_client = sui_sdk_client
+            .as_ref()
+            .ok_or_else(|| Error::Internal("Sui SDK client not initialized".to_string()))
+            .extend()?;
+
+        let exec_arg = if let Some(TransactionMeta {
+            sender,
+            gas_price,
+            gas_objects, // TODO: How do we make this a mock coin when None?
```

Comment:
Looking at how dev inspect and dry run are done today in json rpc, and assuming that we are fine with the current behavior, I think gas object should always be provided when dry running, and not provided (or ignored) when dev inspecting. Does this sound right? And right now in dev inspect, we just create a random gas coin and use it for execution.

Reply:
The behaviour of `dryRunTransactionBlock` is different from the existing dev-inspect and dry-run, mainly because there are edge cases in feature support that sit somewhere between dev-inspect and dry-run that are currently supported by neither but which GraphQL's endpoint will support.

In the case of the gas object, actually both dry run and dev inspect support running without any gas objects supplied.  (Dry running a transaction with no gas object supplied is how gas estimation works).